src/Aspire.Hosting/ContainerResourceBuilderExtensions.cs

```diff
@@ -495,6 +495,131 @@ public static IResourceBuilder<T> WithDockerfile<T>(this IResourceBuilder<T> bui
+    /// <returns>The <see cref="IResourceBuilder{T}"/>.</returns>
+    /// <remarks>
+    /// <para>
+    /// When this method is called, an annotation is added to the <see cref="ContainerResource"/> that specifies the context path
+    /// and a factory function that generates Dockerfile content. The factory is invoked at build time to produce the Dockerfile,
+    /// which is then written to a temporary file and used by the orchestrator to build the container image.
+    /// </para>
+    /// <para>
+    /// The <paramref name="contextPath"/> is relative to the AppHost directory unless it is fully qualified.
+    /// </para>
+    /// <para>
+    /// The factory function is invoked once during the build process to generate the Dockerfile content.
+    /// The output is trusted and not validated.
+    /// </para>
+    /// <example>
+    /// Creates a container called <c>mycontainer</c> with a dynamically generated Dockerfile.
+    /// <code language="csharp">
+    /// var builder = DistributedApplication.CreateBuilder(args);
+    ///
+    /// builder.AddContainer("mycontainer", "myimage")
+    ///        .WithDockerfile("path/to/context", () => 
+    ///        {
+    ///            return "FROM alpine:latest\nRUN echo 'Hello World'";
+    ///        });
+    ///
+    /// builder.Build().Run();
+    /// </code>
+    /// </example>
+    /// </remarks>
+    public static IResourceBuilder<T> WithDockerfile<T>(this IResourceBuilder<T> builder, string contextPath, Func<string> dockerfileFactory, string? stage = null) where T : ContainerResource
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+        ArgumentException.ThrowIfNullOrEmpty(contextPath);
+        ArgumentNullException.ThrowIfNull(dockerfileFactory);
+
+        var fullyQualifiedContextPath = Path.GetFullPath(contextPath, builder.ApplicationBuilder.AppHostDirectory)
+                                           .TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
+
+        // Create a unique temporary Dockerfile path for this resource
+        var tempDockerfilePath = Path.Combine(Path.GetTempPath(), $"Dockerfile.{builder.Resource.Name}.{Guid.NewGuid():N}");
+
+        var imageName = ImageNameGenerator.GenerateImageName(builder);
+        var imageTag = ImageNameGenerator.GenerateImageTag(builder);
+        var annotation = new DockerfileBuildAnnotation(fullyQualifiedContextPath, tempDockerfilePath, stage)
+        {
+            DockerfileFactory = _ => Task.FromResult(dockerfileFactory())
+        };
+
+        return builder.WithAnnotation(annotation, ResourceAnnotationMutationBehavior.Replace)
+                      .WithImageRegistry(registry: null)
+                      .WithImage(imageName)
+                      .WithImageTag(imageTag);
+    }
+
+    /// <summary>
+    /// Causes .NET Aspire to build the specified container image from a Dockerfile generated by an asynchronous factory function.
```

Comment:
Remove .NET Aspire from this. Follow the theme of the other doc comments in this file.